### PR TITLE
UCT/IB/TOOLS: Shrink the size of IB address.

### PR DESCRIPTION
**Pull request by yosefe**
Infiniband specification defines some special values for subnet prefix. If the GID is subnet-local or site-local, we can reduce the address size.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1910/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1287/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1290/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1911/ for details.

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -526,12 +514,22 @@ uct_ud_send_skb_t *uct_ud_ep_prepare_creq(uct_ud_ep_t *ep)
...
+                                   (void*)&creq->conn_req.ep_addr);
+    if (status != UCS_OK) {
+        return NULL;
+    }
+
+    status = uct_ib_iface_get_device_address(&iface->super.super.super,
+                                             (uct_device_addr_t*)uct_ud_creq_ib_addr(creq));
+    if (status != UCS_OK) {
+        return NULL;
+    }
```
Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_ep.c]: Need to free skb on error or to take address before acquiring skb.

**Comment by alex-mikheev** [comment]
It will be good to have unit test that checks site and global ib address pack/unpack

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1913/ for details.

File: src/uct/ib/ud/base/ud_ep.c (hunk 1 above)
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: Do we really need to free the skb? Since it's just taken from `iface->tx.skb`

Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_ep.c]: you are right. tx.skb holds a prefetched skb. So freeing it actually will be an error. It is still better to take address before skb allocation in order not to rely on uct_ud_ep_get_tx_skb() behaviour.

I will add comments to the function that explains its usage.

Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: i can't take address earlier since i have to provide buffer to pack the address into (which is the skb)..

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1298/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1921/ for details.

**Comment by yosefe** [comment]
@alex-mikheev  pls review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1922/ for details.

**Comment by alex-mikheev** [comment]
:+1:

**Pull request merged**